### PR TITLE
[Model] enable data parallel for Llama4 vision encoder

### PR DESCRIPTION
**Pull request by jennyyyyzhen**
**Summary:**
Llama4 vision encoder in dp8 is ~3x as fast as in tp8, especially when handling a large number of input images (eg. 9 images per request). 
Add an enable_vision_encoder_data_parallel to allow using different parallelism for vision model and language model.

**Unit test**
` pytest tests/models/multimodal/generation/test_common.py -k "llama4"`

**MM Eval**
Baseline TP8
```
lm_eval --model vllm-vlm --model_args pretrained=meta-llama/Llama-4-Scout-17B-16E-Instruct,tensor_parallel_size=8,max_model_len=32768,gpu_memory_utilization=0.9 --tasks chartqa --batch_size auto --apply_chat_template
```

| Tasks |Version|Filter|n-shot|     Metric      |   |Value |   |Stderr|
|-------|------:|------|-----:|-----------------|---|-----:|---|-----:|
|chartqa|      0|none  |     0|anywhere_accuracy|↑  |0.8872|±  |0.0063|
|       |       |none  |     0|exact_match      |↑  |0.6548|±  |0.0095|
|       |       |none  |     0|relaxed_accuracy |↑  |0.8848|±  |0.0064|

DP8
```
lm_eval --model vllm-vlm --model_args pretrained=meta-llama/Llama-4-Scout-17B-16E-Ins
truct,tensor_parallel_size=8,max_model_len=32768,gpu_memory_utilization=0.9,enable_multimodal_encoder_data_parallel=true --tasks chartqa --batch_size auto --apply_chat_template
```

| Tasks |Version|Filter|n-shot|     Metric      |   |Value |   |Stderr|
|-------|------:|------|-----:|-----------------|---|-----:|---|-----:|
|chartqa|      0|none  |     0|anywhere_accuracy|↑  |0.8852|±  |0.0064|
|       |       |none  |     0|exact_match      |↑  |0.6492|±  |0.0095|
|       |       |none  |     0|relaxed_accuracy |↑  |0.8820|±  |0.0065|

**perf result**
```
HF_CHECKPOINT=meta-llama/Llama-4-Scout-17B-16E-Instruct
 python benchmarks/benchmark_serving.py --backend openai-chat --model $HF_CHECKPOINT --dataset-name hf --dataset-path lmarena-ai/VisionArena-Chat --hf-split train --num-prompts 1000 --endpoint /v1/chat/completions --max-concurrency 32 --ignore-eos --seed 0
```

DP8
```
============ Serving Benchmark Result ============
Successful requests:                     1000      
Benchmark duration (s):                  115.99    
Total input tokens:                      87321     
Total generated tokens:                  128000    
Request throughput (req/s):              8.62      
Output token throughput (tok/s):         1103.55   
Total Token throughput (tok/s):          1856.38   
---------------Time to First Token----------------
Mean TTFT (ms):                          641.44    
Median TTFT (ms):                        571.65    
P99 TTFT (ms):                           2443.98   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          23.86     
Median TPOT (ms):                        23.52     
P99 TPOT (ms):                           31.16     
---------------Inter-token Latency----------------
Mean ITL (ms):                           23.69     
Median ITL (ms):                         17.91     
P99 ITL (ms):                            222.95    
==================================================
```
Baseline TP8
```
============ Serving Benchmark Result ============
Successful requests:                     1000      
Benchmark duration (s):                  124.06    
Total input tokens:                      87321     
Total generated tokens:                  128000    
Request throughput (req/s):              8.06      
Output token throughput (tok/s):         1031.74   
Total Token throughput (tok/s):          1735.58   
---------------Time to First Token----------------
Mean TTFT (ms):                          758.51    
Median TTFT (ms):                        734.38    
P99 TTFT (ms):                           1805.42   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          24.95     
Median TPOT (ms):                        24.71     
P99 TPOT (ms):                           30.54     
---------------Inter-token Latency----------------
Mean ITL (ms):                           24.76     
Median ITL (ms):                         17.87     
P99 ITL (ms):                            283.55    
==================================================
```

### CODE REVIEW

**Comment by sarckk** [comment]
looks good to me overall, could you also add MM eval results with DP and TP?

**Review by shinoss — COMMENTED** [review]
this looks good to me overall, could you also add MM eval results with DP and TP?

**Comment by DarkLight1337** [comment]
cc @houseroad

**Review by ywang96 — COMMENTED** [review]
Sorry for the delayed review and thank you for the contribution! Overall and I left some comments!

File: vllm/engine/arg_utils.py [hunk 1]
```diff
@@ -634,6 +637,9 @@ def add_cli_args(parser: FlexibleArgumentParser) -> FlexibleArgumentParser:
                                     **parallel_kwargs["worker_cls"])
         parallel_group.add_argument("--worker-extension-cls",
                                     **parallel_kwargs["worker_extension_cls"])
+        parallel_group.add_argument(
+            "--enable-vision-encoder-data-parallel",
+            **parallel_kwargs["enable_vision_encoder_data_parallel"])
```
[the remark this answers is not in the document]
Comment (houseroad) [reply vllm/engine/arg_utils.py:642]: So we use mm for now, then later if needed, we can split into vision and voice?

**Review by houseroad — COMMENTED** [review]
Looks pretty reasonable. Could you rebase and address @ywang96 's feedback?

File: vllm/multimodal/utils.py [hunk 2]
```diff
@@ -390,3 +393,36 @@ def modality_group_func(mm_input: MultiModalKwargs) -> Union[str, int]:
...
+    """
+
+    num_chunks = image_input.shape[0]
+    mp_world_size = get_tensor_model_parallel_world_size()
+    num_chunks_per_rank = (num_chunks + mp_world_size - 1) // mp_world_size
+    num_padded_chunks = num_chunks_per_rank * mp_world_size - num_chunks
+    pad = (0, ) * (2 * (image_input.dim() - 1)) + (0, num_padded_chunks)
+    image_input_padded = torch.nn.functional.pad(image_input, pad)
+    rank = get_tensor_model_parallel_rank()
+    image_input_per_rank = image_input_padded[rank *
```
Comment (houseroad) [inline vllm/multimodal/utils.py:419]: wondering if we need clone here?

File: vllm/multimodal/utils.py [hunk 3]
```diff
@@ -390,3 +393,36 @@ def modality_group_func(mm_input: MultiModalKwargs) -> Union[str, int]:
     return [
         list(group) for _, group in groupby(mm_inputs, key=modality_group_func)
     ]
+
+
+def run_dp_sharded_vision_model(image_input: torch.Tensor,
```
Comment (houseroad) [inline vllm/multimodal/utils.py:398]: we can add some unittest for function. (good for a follow up PR)

File: vllm/model_executor/models/mllama4.py [hunk 4]
```diff
@@ -818,9 +914,12 @@ def load_weights(self, weights: Iterable[tuple[str,
         assert loaded_language_model_params is not None
         updated_params.update(loaded_language_model_params)
 
+        if self.use_data_parallel:
+            other_weights = self._consolidate_qkv_weights(other_weights)
```
Comment (houseroad) [inline vllm/model_executor/models/mllama4.py:918]: nit: we can add some sanity checks to ensure _consolidate_qkv_weights operate appropriately.

**Review by houseroad — APPROVED** [review]
Looks good to me.

**Comment by houseroad** [comment]
@ywang96 , could you give another pass?

File: vllm/engine/arg_utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ywang96) [reply vllm/engine/arg_utils.py]: Yep!

**Review by ywang96 — APPROVED** [review]
Please address Lu's comment otherwise LGTM!

File: vllm/multimodal/utils.py (hunk 3 above)
Comment (jennyyyyzhen) [reply vllm/multimodal/utils.py:398]: @houseroad follow up PR is here https://github.com/vllm-project/vllm/pull/19103
